File: androidApp/src/main/java/com/badmintontracker/android/localanalysis/AnalysisFiles.kt
package com.badmintontracker.android.localanalysis

import java.io.File

/**
 * What an analysis leaves on disk, and how it is thrown away.
 *
 * The stores a run writes into, named in one place rather than at each of
 * their own call sites: [PlayerTrackStore] and [SkeletonStore] each own a
 * local pair and a cloud pair, and [LocalAnalysisRunner] the copied source,
 * and none of them knew about a deletion. A match removed from the app kept
 * its cut clips - tens of megabytes - and its skeleton for the life of the
 * install, reachable by nothing.
 *
 * The cloud stores are included on the same footing as the local ones:
 * removing the entry already removes its local analysis, and nothing protects
 * a cloud artifact the way it protects a local one - a local track costs half
 * an hour of device time to recreate, a cloud one only a re-download. A match
 * uploaded from a different phone has no local entry here to delete, so its
 * artifacts are never reached by this.
 *
 * The iOS side is `AnalysisFiles.deleteAll`, over the local names only for
 * now; it picks up the cloud names in the task that gives it a cloud path.
 */
object AnalysisFiles {

    /**
     * Two shapes: a directory named for the entry (the clips) and a file named
     * for it with an extension (everything else). Both are tried for every
     * store rather than mapped one to one, so adding a store here is one string
     * and cannot get its shape wrong.
     */
    private val STORES = listOf(
        "player-tracks", "skeletons", "local-clips", "local-sources",
        // Cloud artifacts for this entry, not this device's own analysis.
        // Removing the entry already removes its local analysis, and a cloud
        // artifact has no equivalent reason to survive: it costs a
        // re-download, not half an hour of device time. A match uploaded
        // from a different phone has no local entry to delete, so this never
        // reaches an artifact whose local entry is elsewhere.
        "cloud-tracks", "cloud-skeletons",
    )
    private val SUFFIXES = listOf(".track", ".skel", ".mp4")

    /** Removes everything an analysis produced for one video. */
    fun deleteAll(filesDir: File, entryId: String) {
        for (store in STORES) {
            val base = File(filesDir, store)
            File(base, entryId).deleteRecursively()
            for (suffix in SUFFIXES) File(base, entryId + suffix).delete()
        }
    }
}

File: iosApp/Sources/LocalAnalysis/AnalysisFiles.swift
import Foundation

/// Where an on-device analysis keeps what it produced.
///
/// Android hangs everything off `context.filesDir`. The iOS equivalent is
/// Application Support, not Documents: Documents is where `LocalVideoFiles`
/// puts videos because those are the user's own recordings and belong in a
/// backup and, on a file-sharing build, in Files. A track, a skeleton and a set
/// of cut clips are derived - re-analysing reproduces them - so they belong
/// beside the app's other private state, out of the user's way and out of
/// iCloud's.
///
/// Excluded from backup for the same reason: a 30-minute match's skeleton is
/// 11MB and its clips are tens of megabytes of re-encoded video, and pushing
/// that into a coach's iCloud quota to save re-running an analysis they can
/// re-run is not a trade this app gets to make for them.
enum AnalysisFiles {

    private static let root = "Analysis"

    /// Paths are relative to this and resolved on every use, never persisted:
    /// the app container path changes across updates, which is the same reason
    /// `LocalVideoFiles` persists paths relative to Documents.
    static var directory: URL {
        let support = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        return support.appendingPathComponent(root, isDirectory: true)
    }

    /// Creates a subdirectory of the analysis root, excluding the root from
    /// backup the first time it is made.
    @discardableResult
    static func directory(_ name: String) throws -> URL {
        var base = directory
        let existed = FileManager.default.fileExists(atPath: base.path)
        let url = base.appendingPathComponent(name, isDirectory: true)
        try FileManager.default.createDirectory(at: url, withIntermediateDirectories: true)
        if !existed {
            var values = URLResourceValues()
            values.isExcludedFromBackup = true
            try? base.setResourceValues(values)
        }
        return url
    }

    /// Removes everything an analysis produced for one video.
    ///
    /// Called when the video leaves the phone. Without it a deleted match keeps
    /// its clips - tens of megabytes - and its skeleton for the life of the
    /// install, reachable by nothing.
    static func deleteAll(entryId: String) {
        // The cloud stores are here on the same footing as the local ones:
        // removing the entry already removes its local analysis, and nothing
        // protects a cloud artifact the way it protects a local one - a local
        // track costs half an hour of device time to recreate, a cloud one
        // only a re-download. A match uploaded from a different phone has no
        // local entry here, so this never reaches an artifact whose entry
        // lives elsewhere. Mirrors Android's AnalysisFiles.STORES.
        let stores = [
            "player-tracks", "skeletons", "local-clips", "local-sources",
            "cloud-tracks", "cloud-skeletons",
        ]
        for store in stores {
            let base = directory.appendingPathComponent(store, isDirectory: true)
            // Two shapes: a directory named for the entry (clips) and a file
            // named for it with an extension (the rest). Removing a path that
            // is not there is not an error worth reporting.
            try? FileManager.default.removeItem(at: base.appendingPathComponent(entryId, isDirectory: true))
            for suffix in [".track", ".skel", ".mp4"] {
                try? FileManager.default.removeItem(at: base.appendingPathComponent(entryId + suffix))
            }
        }
    }
}
